File: 3-26/139单词拆分/139.py


from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        self.record = set()
        return self.helper(s, wordDict)

    def helper(self, s, wordDict):
        if not s:
            return True

        if s in self.record:
            return False

        for word in wordDict:
            if len(s) >= len(word) and s[:len(word)] == word:
                res = self.helper(s[len(word):], wordDict)
                if res:
                    return True
                else:
                    self.record.add(s)
        return False

File: 2020-05/5-07/139单词拆分/139.cpp

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        // dp, 状态转移方程： dp[j] = dp[i] && s.substr(i, j - i) in record
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true;
        unordered_set<string> record(wordDict.begin(), wordDict.end());

        for(int i = 0; i < s.size(); ++i)
            for(int j = i + 1; j < dp.size(); ++j)
                if(dp[i] && record.find(s.substr(i, j - i)) != record.end())  // 注意是j-i 而不是j-i+1
                    dp[j] = true;

        return dp[dp.size() - 1];
    }
};